### PR TITLE
add developer environment for vscode

### PR DESCRIPTION
**Pull request by vsoch**
Problem: as a new developer it is hard to setup a local environment
Solution: provide a .devcontainer setup, including ruby, chef resources, Python, and tox, and a brief note about using the environment in the CONTRIBUTING.md

### Description of changes
* Describe *what* you're changing and *why* you're doing these changes.

I'm planning on contributing to this project, and I'd like to have a developer environment to not need to install the dependencies directly to my system. I use VSCode (as I'm sure others do!) and a nice solution is to provide a .devcontainer directory with a container environment with everything I need, and instructions in the CONTRIBUTING.md

### Tests
* No tests were needed.

### References
* https://code.visualstudio.com/docs/devcontainers/containers

### Checklist
- [x] Make sure you are pointing to **the right branch**.
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Check if documentation is impacted by this change. (CONTRIBUTING.md)

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    sudo \
+    python3-pip \
+    python3 && \
+    pip3 install tox && \
+    ln -s /usr/bin/python3 /usr/bin/python && \
+    apt-get clean
+
+ENV CHEF_LICENSE=accept-no-persist
+COPY ./Gemfile /tmp/Gemfile
+COPY ./Gemfile.lock /tmp/Gemfile.lock
```
Comment (enrico-usai) [inline .devcontainer/Dockerfile]: Could you replace:
```
COPY ./Gemfile /tmp/Gemfile
COPY ./Gemfile.lock /tmp/Gemfile.lock
```
with:
```
COPY ./Gemfile* /tmp/
```

In this way the command won't fail if the lock file does not exist.

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    chmod +x chefDownload.sh && \    
+    ./chefDownload.sh -c stable -P cinc-workstation
+
+# Add the group and user that match our ids
+RUN groupadd -g ${USER_GID} ${USERNAME} && \
+    adduser --disabled-password --uid ${USER_UID} --gid ${USER_GID} --gecos "" ${USERNAME} && \
+    echo "${USERNAME} ALL=(ALL) NOPASSWD: ALL" > /etc/sudoers
+USER $USERNAME
+ENV PATH=/opt/chef-workstation/embedded/bin:$PATH
+ENV CHEF_LICENSE=accept-no-persist
```
Comment (enrico-usai) [inline .devcontainer/Dockerfile]: This is already exported in the line 25.

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{
+    "name": "AWS Cookbook Developer Environment",
```
Comment (enrico-usai) [inline .devcontainer/devcontainer.json]: AWS ParallelCluster Cookbook Developer Environment

File: .devcontainer/devcontainer.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    "context": "../",
+
+    "customizations": {
+      "vscode": {
+        "settings": {
+          "terminal.integrated.defaultProfile.linux": "bash",
+
+          // Ensure that Python autocomplete works out of the box
+          "python.autoComplete.extraPaths": [
+            "/usr/local/lib/python3.8/site-packages",
```
Comment (enrico-usai) [inline .devcontainer/devcontainer.json]: minor: useless trailing comma

File: .devcontainer/devcontainer.json [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+          ],
+          "python.analysis.extraPaths": [
+            "/usr/local/lib/python3.8/site-packages",
```
Comment (enrico-usai) [inline .devcontainer/devcontainer.json]: minor: useless trailing comma

File: .devcontainer/devcontainer.json [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+          ]
+        },
+        // Note to Flux Developers! We can add extensions here that you like
+        "extensions": [
+          "ms-python.python", // Python support
+          "GitHub.vscode-pull-request-github", // manage and review PRs
```
Comment (enrico-usai) [inline .devcontainer/devcontainer.json]: minor: useless trailing comma

File: .devcontainer/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,41 @@
+FROM ruby:2.7
+
+LABEL maintainer="Vanessasaurus <@vsoch>"
```
Comment (enrico-usai) [inline .devcontainer/Dockerfile]: Could you change it to "Amazon Web Services" to be aligned with the maintainer of the CLI? https://github.com/aws/aws-parallelcluster/blob/develop/cli/setup.py#L63

**Review by enrico-usai — COMMENTED** [review]
Hi @vsoch!

I tested it and it is really useful, thanks for your contribute.
I left some minor comments.

Thanks
Enrico

**Comment by vsoch** [comment]
Fixed, rebased, and tested locally. Thank you for the review @enrico-usai!

**Comment by enrico-usai** [comment]
Unit tests are failing because of a recent release of pytest-html, let's wait to have a fix from other devs in develop then we'll rebase the pr and will merge it.

**Comment by vsoch** [comment]
Sounds good! Please ping me when you are ready for a rebase.

**Comment by enrico-usai** [comment]
Develop contains the fix, you can rebase your patch, thanks!

**Comment by vsoch** [comment]
> Develop contains the fix, you can rebase your patch, thanks!

All set - fingers crossed!

**Review by enrico-usai — APPROVED** [review]

**Comment by vsoch** [comment]
Woohoo - thank you @enrico-usai ! How would you suggest someone new to this setup start to work on adding to it? Any tips for a new cookbook developer?

**Comment by enrico-usai** [comment]
I think the best way to start working in the cookbook is to look at kitchen and inspec tests (see README) and try to write new kitchen/inspec tests to cover the uncovered recipes/resources. This permits to gain confidence in the code and add robustness to the code itself.

**Pull request merged**